* draw right region narrow
* check for exact match
* bad primers BEDs
** CCDC106+1 - outside of target
** DPP3+1 - doesn't cross intron
* CERNA1+1 - some of isPCr showed going into retained intron
* add TM to primer to spread sheet ; want highest
* Another suggestion if possible for better visualisation
We should find a way that we can easy see if delta_G is ok
should be weaker (more positive) than -9.0 kcal/mole
* Ordered primers by  score
* rna annotates doesn't work right
* renamed ppair_id
* delete empty columns in spec
* primer_design -> design_pairs
* rename to indicated if they are features_5p, features_3p are genome or transcriptome order
* created primers that went beyond 3' end of region.  This maybe something primer3 does
see test_design_primers_guts.py test_FBXL16(request, genome_data, wtc11_targets_specs_set1):
* include seq_args and global_args in Primer3Results and dump from there,
primer3_dump_args should not have to recalculate
* bounds() name not so clear
* '+' is a bit of a confusing name separator, maybe CCDC106_1_pp1
